### PR TITLE
Improve client b1 to implement 1

### LINKED ISSUE #37: There should be a method for each Apollo send type
**Describe the feature or change you'd like**
There is currently a method that takes care of every type. The type can be specified using a string. Each type (query, mutation) should get its own method.

### LINKED ISSUE #38: There should be less interfaces
**Describe the feature or change you'd like**
Currently there are plenty of interfaces which could be removed if Apollo endpoint could handle types.
Therefore type parameter handling should be added to Apollo in order to have better type handling and to reduce a interface overhead.

**Additional context**
- Ref: https://www.typescriptlang.org/docs/handbook/generics.html

### LINKED ISSUE #39: There should be a way to perform custom SNEK tasks
**Describe the feature or change you'd like**
It should be possible to send a specific query or mutation to the seek-engine which are not defined in the task template. This is possible by implementing a custom task functionality.

### LINKED ISSUE #40: Change client apollo endpoint notation
**Describe the feature or change you'd like**
Currently the graphql endpoint is accessible via `client.endpoint`. This notation does not represent the functionality. Therefore it should be changed to `client.gql`
When a client has multiple endpoints what name would they have?

Like this?: `client.endpoint1
client.endpoint2
client.scraper
`

That would create a lot of confusion and should therefore be changed

**Additional context**
Ref: https://github.com/snek-at/client/blob/implement-1/src/index.ts

### PR DESCRIPTION
**Pull request by schettn**
**Checklist:**

 
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [x] Have you updated or added documentation for the change?
 - [x] Have you tested your changes with successful results?

**Type of Changes:**

 
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] New feature (non-breaking change which adds functionality)
 - [ ] Documentation (non-breaking change which adds documentation)
 - [x] Breaking change (fix or feature that would cause existing functionality to change)

**What is the current behavior? (link to any open issues here)**
- Currently there is one method which handles all Apollo types. #37 
- Currently there are plenty of interfaces. #38 
- Currently it is not possible to perform a custom session aware task. #39 
- Currently the clients Apollo endpoint notation is misleading. #40 

**What is the new behavior (if this is a feature change)?**
- Now there are separate methods for each Apollo type. Closes #37 
- Now the number of interfaces has been removed by adding type parameter handling to Apollo endpoint. Closes #38 
- Now there it is possible to perform custom session aware tasks. Closes #39 
- Now the notation of the Apollo endpoint of clients has been adjusted. Currently the old version is still present but will be removed in a upcoming release. #40

**Other information**:
- Ref: 🐍

### CODE REVIEW

File: src/endpoints/apollo.ts [hunk 1]
```diff
@@ -75,52 +78,75 @@ class Apollo implements ApolloEndpoint {
...
+  ) {
+    if (type === "query") {
+      return this.sendQuery(data, variables, headers);
+    } else if (type === "mutation") {
+      return this.sendMutation(data, variables, headers);
+    } else {
+      console.warn(
+        "No query type specified! Please re-check." +
+          "Selecting type 'query' as default"
+      );
```
Comment (pinterid) [inline src/endpoints/apollo.ts:97]: Missing newline.

File: src/endpoints/index.ts [hunk 2]
```diff
@@ -5,6 +5,12 @@
 // Contains the interface for gql queries, mutations and subscriptions
 import { DocumentNode } from "graphql";
 //#endregion
+import { FetchResult } from "apollo-link";
+import { ApolloQueryResult } from "apollo-client";
+
+export type ApolloResult<T> =
```
Comment (pinterid) [inline src/endpoints/index.ts]: Put export to the bottom of the file.

File: src/session/sessions.ts [hunk 3]
```diff
@@ -168,16 +177,16 @@ class SnekSession extends Session {
...
-    return this.ep.send("query", data, variables, headers);
+    return this.ep.sendQuery(data, variables, headers);
   }
 
   /**
    * Begin session.
    *
-   * @param {string} user A User defined by username and password
-   * @returns {Promise<UserData>} A UserData object
+   * @param {string} user A User defined by username and password.
```
Comment (pinterid) [inline src/session/sessions.ts]: Remove dot.

File: src/session/sessions.ts [hunk 4]
```diff
@@ -168,16 +177,16 @@ class SnekSession extends Session {
...
+   * @returns {Promise<any>} A UserData object.
```
Comment (pinterid) [inline src/session/sessions.ts]: Remove dot.

File: src/templates/snek/gql/tasks/auth.ts [hunk 5]
```diff
@@ -1,101 +1,70 @@
...
 interface RevokeData {
-  revoked: string;
+  revoke: {
+    revoked: string;
+  };
 }
 //#endregion
 
 //#region > Classes
 /** @class A set of session aware Tasks */
```
Comment (pinterid) [inline src/templates/snek/gql/tasks/auth.ts]: Write "Tasks" in lower case.

File: src/templates/snek/gql/tasks/auth.ts [hunk 6]
```diff
@@ -1,101 +1,70 @@
...
-
-    this.template = session.template.snekGql;
-  }
+  constructor(private parent: SnekTasks) {}
 
   /**
    * Anonymous login.
    *
-   * @returns {Promise<AuthResponse>} A JWT token
+   * @returns {Promise<ApolloResult<AuthData>>} A JWT token.
```
Comment (pinterid) [inline src/templates/snek/gql/tasks/auth.ts]: Remove dot and is "token" needed in this case?

File: src/templates/snek/gql/tasks/auth.ts [hunk 7]
```diff
@@ -109,64 +78,55 @@ class SnekGqlAuthTasks extends TaskError {
    * User login.
    *
    * @param {string} user A User defined by username and password
-   * @returns {Promise<AuthData>} A JWT token
    * @description Authenticates a real user to obtain a JWT
+   * @returns {Promise<ApolloResult<AuthData>>} A JWT token
```
Comment (pinterid) [inline src/templates/snek/gql/tasks/auth.ts]: Is "token" required here?

File: src/templates/snek/gql/tasks/auth.ts [hunk 8]
```diff
@@ -109,64 +78,55 @@ class SnekGqlAuthTasks extends TaskError {
...
-
     return response;
   }
 
   /**
    * Refresh a token.
    *
    * @param {string} user A User defined by username and password
-   * @returns {Promise<RefreshResponse>} A JWT token
+   * @returns {Promise<ApolloResult<RefreshData>>} A JWT token
```
Comment (pinterid) [inline src/templates/snek/gql/tasks/auth.ts]: Is "token" required here?

File: src/templates/snek/gql/tasks/general.ts [hunk 9]
```diff
@@ -1,99 +1,76 @@
...
- */
-interface IGitlabServerResponse extends IResponse {
-  data: { page: GitlabServerData };
-}
-
-/**
- * @interface GitlabServerData defines the structure of the specific data a
- *                             GitlabServerResponse contains.
+ * @interface GitlabServerData defines the structure of the Gitlab server result
+ *                             data
```
Comment (pinterid) [inline src/templates/snek/gql/tasks/general.ts]: Missing dot.

File: src/templates/snek/gql/tasks/general.ts [hunk 10]
```diff
@@ -1,99 +1,76 @@
...
-class SnekGqlGeneralTasks extends TaskError {
-  public template: ISnekGqlTemplate;
-
+class SnekGqlGeneralTasks {
   /**
    * @constructor
    * @author Nico Schett <contact@schett.net>
-   * @param {string} session A session for the tasks
+   * @param {string} parent The parent task
+   * @param {string} template A template set containing mutation and queries
```
Comment (pinterid) [inline src/templates/snek/gql/tasks/general.ts]: Use the plural of "mutation"

File: src/templates/snek/gql/tasks/index.ts [hunk 11]
```diff
@@ -38,15 +38,40 @@ class SnekTasks {
...
    */
   constructor(session: SnekSession) {
-    this.general = new SnekGqlGeneralTasks(session);
-    this.auth = new SnekGqlAuthTasks(session);
-    this.user = new SnekGqlUserTasks(session);
+    super(session);
+
+    this.template = session.template.snekGql;
+
+    /* init tasks */
```
Comment (pinterid) [inline src/templates/snek/gql/tasks/index.ts]: Change "init" to "Init"

File: src/templates/snek/gql/tasks/index.ts [hunk 12]
```diff
@@ -38,15 +38,40 @@ class SnekTasks {
...
+
+    if (type === "query") {
+      response = await this.session.ep.sendQuery<T>(query, variables);
+    } else if (type === "mutation") {
+      response = await this.session.ep.sendMutation<T>(query, variables);
+    } else {
+      console.warn(
+        "No query type specified! Please re-check." +
+          "Selecting type 'query' as default"
+      );
```
Comment (pinterid) [inline src/templates/snek/gql/tasks/index.ts:62]: Missing newline.

File: src/templates/snek/gql/tasks/user.ts [hunk 13]
```diff
@@ -73,116 +43,98 @@ interface ProfileData {
...
- */
-interface IWhoamiResponse extends IResponse {
-  data: WhoamiData;
-}
-
-/**
- * @interface WhoamiData defines the structure of the specific data a
- *                       whoami response contains.
+ * @interface WhoamiData defines the structure of the whoami result
+ *                        data.
```
Comment (pinterid) [inline src/templates/snek/gql/tasks/user.ts:47]: Remove space before "data".

File: src/templates/snek/gql/tasks/user.ts [hunk 14]
```diff
@@ -73,116 +43,98 @@ interface ProfileData {
...
-class SnekGqlUserTasks extends TaskError {
-  public template: ISnekGqlTemplate;
-
+class SnekGqlUserTasks {
   /**
    * @constructor
    * @author Nico Schett <contact@schett.net>
-   * @param {string} session A session for the tasks
+   * @param {string} parent The parent task
+   * @param {string} template A template set containing mutation and queries
```
Comment (pinterid) [inline src/templates/snek/gql/tasks/user.ts]: Use the plural of "mutation"

File: src/templates/snek/gql/tasks/user.ts [hunk 15]
```diff
@@ -73,116 +43,98 @@ interface ProfileData {
...
-    this.template = session.template.snekGql;
-  }
+  constructor(private parent: SnekTasks) {}
 
   /**
    * Register a user.
    *
    * @param {object} values Registration parameters like username, email,...
-   * @returns {Promise<RegistrationResponse>} A JWT token
+   * @returns {Promise<ApolloResult<RegistrationData>>} A JWT token
```
Comment (pinterid) [inline src/templates/snek/gql/tasks/user.ts]: Is "token" required?

File: src/templates/snek/gql/tasks/user.ts [hunk 16]
```diff
@@ -73,116 +43,98 @@ interface ProfileData {
...
-
     return response;
   }
 
   /**
    * Cache a user.
    *
    * @param {string} platformData A serialized JSON object to be cached
-   * @returns {Promise<CacheResponse>} A JWT token
+   * @returns {Promise<ApolloResult<CacheData>>} A JWT token
```
Comment (pinterid) [inline src/templates/snek/gql/tasks/user.ts]: Is "token" required?

**Review by pinterid — CHANGES_REQUESTED** [review]
Looks good after some changes.

File: src/templates/snek/gql/tasks/general.ts (hunk 10 above)
Comment (schettn) [reply src/templates/snek/gql/tasks/general.ts]: Tis param should be removed totally.

File: src/templates/snek/gql/tasks/user.ts (hunk 14 above)
Comment (schettn) [reply src/templates/snek/gql/tasks/user.ts]: This should be removed.

**Review by Aichnerc — APPROVED** [review]
LGTM 🎁

**Review by pinterid — APPROVED** [review]
Looks good to merge.

**Pull request merged**